test/parallel/test-error-codes-impl-doc.js

```diff
@@ -0,0 +1,158 @@
+'use strict';
+
+// Flags: --expose-internals
+
+// Note: this script reads Node.js sources and docs, and expects to be run on
+// the very same Node.js version as the source and doc tree is.
+
+require('../common');
+const assert = require('assert');
+const errors = require('internal/errors').codes;
+const fs = require('fs');
+const path = require('path');
+const natives = process.binding('natives');
+
+// --report prints only the list of failed checks and does not assert
+const report = process.argv[2] === '--report';
+const results = [];
+function check(ok, type, name, reason) {
+  const label = `${type}: ${name} - ${reason}`;
+  if (report) {
+    if (!ok) results.push(label);
+  } else {
+    assert.ok(ok, label);
+  }
+}
+
+const set = {
+  all: new Set(),
+  js: new Set(), // Error codes found in natives.
+  doc: new Set(), // Error codes mentioned in `doc/api/*.md`.
+  cpp: new Set(), // Error codes found in `src/node_errors.h`.
+  noargs: new Set(), // Subset of `js` which is constructed without arguments.
+  // `set.manual` contains errors manually created from js and not defined in
+  // internals/errors. That is not scanned now, update this list if required.
+  manual: new Set(['ERR_HTTP2_ERROR', 'ERR_UNKNOWN_BUILTIN_MODULE']),
+  // `set.examples` contains errors mentioned in documentation purely for
+  // demonstration purposes. These errors are not required to be present.
+  examples: new Set(['ERR_ERROR_1'])
+};
+
+const root = path.join(__dirname, '..', '..');
+
+// File containing error definitions
+const jsdata = natives['internal/errors'];
+// File containing error documentation
+let docdata = fs.readFileSync(path.join(root, 'doc/api/errors.md'), 'utf-8')
+docdata = docdata.replace(/## Legacy [\s\S]*/, '');
+// File containing cpp-side errors
+const cppdata = fs.readFileSync(path.join(root, 'src/node_errors.h'), 'utf-8');
+// Directory containing documentation
+const docdir = path.join(root, 'doc/api/');
+
+function addSource(source, type) {
+  // eslint-disable-next-line node-core/no-unescaped-regexp-dot
+  const re = /(.)?(ERR_[A-Z0-9_]+)(..)?/g;
+  let match;
+  while (match = re.exec(source)) {
+    if (match[1] === '_') continue; // does not start with ERR_
+    const name = match[2];
+    if (type === 'doc' && set.examples.has(name)) continue; // is an example
+    set.all.add(name);
+    set[type].add(name);
+    if (type === 'js' && match[3] === '()') {
+      // Is directly called without arguments from js, we will try that
+      set.noargs.add(name);
+    }
+  }
+}
+
+// Add all errors from JS natives
+for (const file of Object.keys(natives)) addSource(natives[file], 'js');
+// Add all errors from src/node_errors.h
+addSource(cppdata, 'cpp');
+// Add all errors from doc/api/*.md files
+for (const file of fs.readdirSync(docdir)) {
+  if (!file.endsWith('.md')) continue;
+  let data = fs.readFileSync(path.join(docdir, file), 'utf-8');
+  if (file === 'errors.md') data = data.replace(/## Legacy [\s\S]*/, '');
```

Comment:
Should `docdata` maybe just be read here? That way the file must only be read once instead of twice.